dask/array/core.py

```diff
@@ -1938,14 +1997,85 @@ def normalize_chunks(chunks, shape=None):
             raise ValueError("Input array has %d dimensions but the supplied "
                              "chunks has only %d dimensions" %
                              (len(shape), len(chunks)))
-        if not all(c == s or (math.isnan(c) and math.isnan(s))
+        if not all(c == s or (math.isnan(c) or math.isnan(s))
                    for c, s in zip(map(sum, chunks), shape)):
             raise ValueError("Chunks do not add up to shape. "
                              "Got chunks=%s, shape=%s" % (chunks, shape))
 
     return tuple(tuple(int(x) if not math.isnan(x) else x for x in c) for c in chunks)
 
 
+def auto_chunks(chunks, shape, limit, dtype, previous_chunks=None):
+    """ Determine automatic chunks
+
+    See also
+    --------
+    normalize_chunks: for full docstring and parameters
+    """
+    if limit is None:
+        limit = config.get('array.chunk-size')
+        if isinstance(limit, str):
+            limit = parse_bytes(limit)
+
+    if dtype is None:
+        raise TypeError("DType must be known for auto-chunking")
+    if dtype.hasobject:
+        raise NotImplementedError("Can not use auto rechunking with object dtype")
+    limit = max(1, limit // dtype.itemsize)
+    chunks = list(chunks)
+
+    autos = {i for i, c in enumerate(chunks) if c == 'auto'}
+    if not autos:
+        return tuple(chunks)
+
+    largest_block = np.prod([cs if isinstance(cs, Number) else max(cs)
+                             for cs in chunks if cs != 'auto'])
+
+    if previous_chunks:
+        result = {a: np.median(previous_chunks[a]) for a in autos}
```

Comment:
median may commonly work, but I could imagine it being far from optimal sometimes. This should at least be in the docstring near "inspiration".

Reply:
I agree that we should document this.  

Have you run into cases where median would lead to pathalogical results?  I would be curious if there is a good pragmatic fail case here.

Reply:
`(1, 1, 1000)` ? Of course, it's not obvious what would be the right thing to do here.

Reply:
Could imagine cases where `max` might be better. Namely mostly uniform chunks that have broken down a bit to occasional/localized splitting of chunks.